src/coord/src/coord.rs

```diff
@@ -3348,6 +3348,29 @@ impl<S: Append + 'static> Coordinator<S> {
         Ok(ExecuteResponse::DroppedRole)
     }
 
+    async fn drop_replica(
+        &mut self,
+        instance_id: ComputeInstanceId,
+        replica_id: ReplicaId,
+        replica_config: ConcreteComputeInstanceReplicaConfig,
+    ) -> Result<(), anyhow::Error> {
+        if let Some(metadata) = self.transient_replica_metadata.remove(&replica_id) {
+            let table = self
+                .catalog
+                .state()
+                .resolve_builtin_table(&MZ_CLUSTER_REPLICA_HEARTBEATS);
+            let retraction = BuiltinTableUpdate {
+                id: table,
+                row: metadata.as_row(replica_id),
+                diff: -1,
+            };
+            self.send_builtin_table_updates(vec![retraction]).await;
+        }
+        self.dataflow_client
+            .drop_replica(instance_id, replica_id, replica_config)
+            .await
+    }
```

Comment:
I was wondering for a bit if we could have a race where we receive another heartbeat after we deleted the replica ID from `transient_replica_metadata`, causing another insert for a replica that we already consider deleted. I assume that is not be possible because the coordinator does not handle any compute commands between deleting from the hashmap and calling `dataflow_client.drop_replica`, and after that returns no commands from that replica are received anymore?

Reply:
I think that's correct.

However, we can make it more defensive by including a tombstone value in the hashmap rather than `None`.